File: src/focalnet/crop.py
"""Aspect-ratio crop selection using all the importance mass."""

import math
from dataclasses import asdict, dataclass

import numpy as np


def clean_map(heatmap: np.ndarray) -> np.ndarray:
    values = np.asarray(heatmap, dtype=np.float64)
    if values.ndim != 2 or not values.size:
        raise ValueError("Expected a nonempty 2D heatmap")
    return np.where(np.isfinite(values) & (values > 0), values, 0)


def focal_point(heatmap: np.ndarray) -> dict[str, float]:
    values = clean_map(heatmap)
    total = values.sum()
    if total == 0:
        return {"x": 0.5, "y": 0.5}
    h, w = values.shape
    return {
        "x": float(values.sum(axis=0) @ ((np.arange(w) + 0.5) / w) / total),
        "y": float(values.sum(axis=1) @ ((np.arange(h) + 0.5) / h) / total),
    }


@dataclass(frozen=True)
class Crop:
    left: int
    top: int
    width: int
    height: int
    retained_importance: float

    def to_dict(self) -> dict:
        return asdict(self)


def _best_offset(mass: np.ndarray, extent: int, window: int) -> tuple[int, float]:
    """Exact search over integer offsets for a piecewise-uniform map density.

    A maximal-area crop spans one whole image axis, reducing the 2D integral
    to a 1D sliding window. Its score changes slope only at heatmap cell edges
    or an edge minus the window width. Evaluate neighboring integers at each
    breakpoint, plus the centered position, to avoid an image-sized scan.
    """
    edges = np.linspace(0, extent, len(mass) + 1)
    cumulative = np.concatenate(([0.0], np.cumsum(mass)))
    breaks = np.concatenate((edges, edges - window, [(extent - window) / 2]))
    candidates = np.unique(
        np.clip(np.concatenate((np.floor(breaks), np.ceil(breaks))), 0, extent - window).astype(int)
    )
    scores = np.interp(candidates + window, edges, cumulative) - np.interp(
        candidates, edges, cumulative
    )
    best = scores.max()
    ties = np.flatnonzero(np.isclose(scores, best, rtol=1e-10, atol=0))
    center = (extent - window) / 2
    index = ties[np.argmin(np.abs(candidates[ties] - center))]
    return int(candidates[index]), float(scores[index])


def solve_crop(
    heatmap: np.ndarray, image_width: int, image_height: int, aspect_ratio: float
) -> Crop:
    """Largest inscribed crop at the requested ratio, rounded to integer pixels.

    Chooses the position retaining the most importance. Equal scores prefer
    the image center. A zero map produces a centered crop and zero retention.
    This optimizes retention, not aesthetics or face-box containment.
    """
    if min(image_width, image_height) <= 0:
        raise ValueError("Image dimensions must be positive")
    if not math.isfinite(aspect_ratio) or aspect_ratio <= 0:
        raise ValueError("Aspect ratio must be finite and positive")
    values = clean_map(heatmap)
    width, height = image_width, image_height
    if width / height > aspect_ratio:
        width = max(1, min(width, int(height * aspect_ratio + 0.5)))
    else:
        height = max(1, min(height, int(width / aspect_ratio + 0.5)))
    if width < image_width:
        left, retained = _best_offset(values.sum(axis=0), image_width, width)
        top = 0
    else:
        top, retained = _best_offset(values.sum(axis=1), image_height, height)
        left = 0
    total = float(values.sum())
    return Crop(left, top, width, height, float(np.clip(retained / total, 0, 1)) if total else 0)


def score_crop(heatmap: np.ndarray, crop: Crop, image_width: int, image_height: int) -> float:
    """Fractional-cell integration of a crop against an independent reference map."""
    values = clean_map(heatmap)
    total = values.sum()
    if total == 0:
        return 0.0
    rows, columns = values.shape
    xs = np.linspace(0, image_width, columns + 1)
    ys = np.linspace(0, image_height, rows + 1)
    wx = np.maximum(
        0, np.minimum(xs[1:], crop.left + crop.width) - np.maximum(xs[:-1], crop.left)
    ) / (image_width / columns)
    wy = np.maximum(
        0, np.minimum(ys[1:], crop.top + crop.height) - np.maximum(ys[:-1], crop.top)
    ) / (image_height / rows)
    return float(np.clip(wy @ values @ wx / total, 0, 1))

File: web/src/crop.ts
import { Mat, interp, isClose, linspace } from "./matrix";

export type Gravity = { x: number; y: number };

export type Crop = {
  left: number;
  top: number;
  width: number;
  height: number;
  retained_importance: number;
};

export function cleanMap(heatmap: Mat): Mat {
  const cleaned = Mat.zeros(heatmap.rows, heatmap.cols);
  for (let index = 0; index < heatmap.data.length; index += 1) {
    const value = heatmap.data[index];
    cleaned.data[index] = Number.isFinite(value) && value > 0 ? value : 0;
  }
  return cleaned;
}

export function focalPoint(heatmap: Mat): Gravity {
  const values = cleanMap(heatmap);
  const total = values.sum();
  if (total === 0) {
    return { x: 0.5, y: 0.5 };
  }
  const { rows, cols } = values;
  let x = 0;
  let y = 0;
  for (let col = 0; col < cols; col += 1) {
    let mass = 0;
    for (let row = 0; row < rows; row += 1) {
      mass += values.at(row, col);
    }
    x += mass * ((col + 0.5) / cols);
  }
  for (let row = 0; row < rows; row += 1) {
    let mass = 0;
    for (let col = 0; col < cols; col += 1) {
      mass += values.at(row, col);
    }
    y += mass * ((row + 0.5) / rows);
  }
  return { x: x / total, y: y / total };
}

function bestOffset(mass: Float64Array, extent: number, window: number): [number, number] {
  const edges = linspace(0, extent, mass.length + 1);
  const cumulative = new Float64Array(mass.length + 1);
  for (let index = 0; index < mass.length; index += 1) {
    cumulative[index + 1] = cumulative[index] + mass[index];
  }
  const center = (extent - window) / 2;
  const unique = new Set<number>();
  const consider = (value: number) => {
    const clipped = Math.min(Math.max(value, 0), extent - window);
    unique.add(clipped);
  };
  for (const edge of edges) {
    consider(Math.floor(edge));
    consider(Math.ceil(edge));
    consider(Math.floor(edge - window));
    consider(Math.ceil(edge - window));
  }
  consider(Math.floor(center));
  consider(Math.ceil(center));
  const candidates = [...unique].sort((left, right) => left - right);
  const scores = candidates.map(
    (offset) => interp(offset + window, edges, cumulative) - interp(offset, edges, cumulative),
  );
  const best = Math.max(...scores);
  let chosen = candidates[0];
  let closest = Infinity;
  for (let index = 0; index < candidates.length; index += 1) {
    if (!isClose(scores[index], best, 1e-10, 0)) {
      continue;
    }
    const distance = Math.abs(candidates[index] - center);
    if (distance < closest) {
      closest = distance;
      chosen = candidates[index];
    }
  }
  return [chosen, best];
}

function halfUp(value: number): number {
  return Math.trunc(value + 0.5);
}

export function solveCrop(
  heatmap: Mat,
  imageWidth: number,
  imageHeight: number,
  aspectRatio: number,
): Crop {
  if (Math.min(imageWidth, imageHeight) <= 0) {
    throw new Error("Image dimensions must be positive");
  }
  if (!Number.isFinite(aspectRatio) || aspectRatio <= 0) {
    throw new Error("Aspect ratio must be finite and positive");
  }
  const values = cleanMap(heatmap);
  let width = imageWidth;
  let height = imageHeight;
  if (width / height > aspectRatio) {
    width = Math.max(1, Math.min(width, halfUp(height * aspectRatio)));
  } else {
    height = Math.max(1, Math.min(height, halfUp(width / aspectRatio)));
  }
  let left = 0;
  let top = 0;
  let retained = 0;
  if (width < imageWidth) {
    const columnMass = new Float64Array(values.cols);
    for (let row = 0; row < values.rows; row += 1) {
      for (let col = 0; col < values.cols; col += 1) {
        columnMass[col] += values.at(row, col);
      }
    }
    [left, retained] = bestOffset(columnMass, imageWidth, width);
  } else {
    const rowMass = new Float64Array(values.rows);
    for (let row = 0; row < values.rows; row += 1) {
      for (let col = 0; col < values.cols; col += 1) {
        rowMass[row] += values.at(row, col);
      }
    }
    [top, retained] = bestOffset(rowMass, imageHeight, height);
  }
  const total = values.sum();
  const fraction = total ? Math.min(1, Math.max(0, retained / total)) : 0;
  return { left, top, width, height, retained_importance: fraction };
}

export function scoreCrop(
  heatmap: Mat,
  crop: Crop,
  imageWidth: number,
  imageHeight: number,
): number {
  const values = cleanMap(heatmap);
  const total = values.sum();
  if (total === 0) {
    return 0;
  }
  const xs = linspace(0, imageWidth, values.cols + 1);
  const ys = linspace(0, imageHeight, values.rows + 1);
  const cellWidth = imageWidth / values.cols;
  const cellHeight = imageHeight / values.rows;
  const weightsX = new Float64Array(values.cols);
  const weightsY = new Float64Array(values.rows);
  const right = crop.left + crop.width;
  const bottom = crop.top + crop.height;
  for (let col = 0; col < values.cols; col += 1) {
    weightsX[col] =
      Math.max(0, Math.min(xs[col + 1], right) - Math.max(xs[col], crop.left)) / cellWidth;
  }
  for (let row = 0; row < values.rows; row += 1) {
    weightsY[row] =
      Math.max(0, Math.min(ys[row + 1], bottom) - Math.max(ys[row], crop.top)) / cellHeight;
  }
  let retained = 0;
  for (let row = 0; row < values.rows; row += 1) {
    if (weightsY[row] === 0) {
      continue;
    }
    let rowMass = 0;
    for (let col = 0; col < values.cols; col += 1) {
      rowMass += values.at(row, col) * weightsX[col];
    }
    retained += weightsY[row] * rowMass;
  }
  return Math.min(1, Math.max(0, retained / total));
}
